hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/web/WebHdfsFileSystem.java

```diff
@@ -2207,10 +2206,8 @@ public boolean hasPathCapability(final Path path, final String capability)
       throws IOException {
     // qualify the path to make sure that it refers to the current FS.
     final Path p = makeQualified(path);
-    Optional<Boolean> cap = DfsPathCapabilities.hasPathCapability(p,
-        capability);
-    if (cap.isPresent()) {
-      return cap.get();
+    if (DfsPathCapabilities.hasPathCapability(p, capability)) {
```

Comment:
what if `validatePathCapabilityArgs(path, capability)` in DfsPathCapablity.hasPathCapablity gives `CommonPathCapabilities.FS_SYMLINKS`. Now, `FileSystem.areSymlinksEnabled()` can be true or false. 
In earlier code, if `FileSystem.areSymlinksEnabled()` is false, we would retrn from old-line 2213 `cap.get()`. But now, it will go ahead and invoke `super.hasPathCapability(p, capability);`

Reply:
Unfortunately, findbugs does not like `Boolean`.  Let me revert the previous change and adds a test to ensure the correctness.

Reply:
Hi. I understand that at present making call to `FileSystem.hasPathCapability()` is not going to make any issue as it checks for symlinks. But what if there is more logic added in the `FileSystem.hasPathCapability()` in future which expects the child classes to call the super method only in required cases.
My thought is that what ever is in `FileSystem.hasPathCapablity` should be abstract to any dev reading `hasPathCapablity` in `DistributedFileSystem` and `WebHdfsFileSystem`, and super.hasPathCapablity should be called only when required. if `Optional<Boolean> cap` gets true/false, it should return from there and not go to the super method. In my view, super method should only get invoked in case given capability doesn't match any of the https://github.com/apache/hadoop/blob/b9fa5e0182c19adc4ff4cd2d9265a36ce9913178/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/client/DfsPathCapabilities.java#L45-L57.
I feel having an enum which can say TRUE, FALSE, UNKNOWN would be great. And the super method getting called only if `DfsPathCapablities.hasPathCapablity` gives UNKNOWN.

Thanks.

Reply:
@saxenapranav , we cannot fix a potential "future" problem.  Since there is a unit test, such future change will be detected by the test.  At that time, we may think about how to solve the problem such as adding the enum.

Reply:
I am completely understanding your point. But my only difference is that if something is returned from `DfsPathCapabilities.hasPathCapability` should be returned back to caller of `DistributedFileSystem.hasPathCapablity), without going to super's method.

I understand that super.hasPathCapability call doesn't pose any issue as of now and we can see for Enum in case more complications are added in future. I have a suggestion in the tests https://github.com/szetszwo/hadoop/commit/6ddc4f245b2f234bc5da5b3f02c1882ecc5fdaec. In this commit it has added asserts on the conditions used in `FileSystem.hasPathCapablity` Would be awesome if you can pick my commit in your pr.

Thanks.

Reply:
@saxenapranav , I am fine to merge your test code.  Thanks.  Is there a way to do it in Github?  Or I have to download it and merge it manually?

Reply:
Thank you so much. I have opened pr on your fork: https://github.com/szetszwo/hadoop/pull/2/files. Thanks.